subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/IvyDependencyMetadata.java

```diff
@@ -101,6 +102,15 @@ public String getDynamicConstraintVersion() {
 
     @Override
     public Set<ConfigurationMetadata> selectConfigurations(ImmutableAttributes consumerAttributes, ComponentResolveMetadata fromComponent, ConfigurationMetadata fromConfiguration, ComponentResolveMetadata targetComponent, AttributesSchemaInternal consumerSchema) {
+        if (!targetComponent.getVariantsForGraphTraversal().isEmpty()) {
+            // This condition shouldn't be here, and attribute matching should always be applied when the target has variants
+            // however, the schemas and metadata implementations are not yet set up for this, so skip this unless:
+            // - the consumer has asked for something specific (by providing attributes), as the other metadata types are broken for the 'use defaults' case
+            // - or the target is a component from a Ivy repo as we can assume this is well behaved
+            if (!consumerAttributes.isEmpty() || targetComponent instanceof IvyModuleResolveMetadata) {
+                return ImmutableSet.of(selectConfigurationUsingAttributeMatching(consumerAttributes, targetComponent, consumerSchema));
+            }
+        }
```

Comment:
It would be great to pull this out so that we can avoid the duplication with `MavenDependencyMetadata`

Reply:
I'm not sure how to handle this, and actually where it makes sense. The comment doesn't really help me either. I added this when chatting with @jjohannes , but I don't think we have code coverage for that. Any help in setting up such a test would be helpful.

Reply:
We should be able to test this by tests with transitive dependencies. And I think it will fail if you have 

`project --depends-on--> module-from-maven --depends-on--> module-from-ivy`

Because of the `targetComponent instanceof ...ModuleResolveMetadata`, which I _think_ should be removed from both Ivy and Maven. And then, the check `!targetComponent.getVariantsForGraphTraversal().isEmpty() && !consumerAttributes.isEmpty()` should go into the super class.

Reply:
@melix Will you look into adding some test coverage for this case? It would be great to fully understand the reason for this check.

It seems like original check for `instanceof MavenModuleResolveMetadata` was required before we added equivalent support for Ivy. At the very least, I think we should change the check to be `instanceof ModuleComponentResolveMetadata`, which will then be identical for Ivy and Maven.

Reply:
Yes, this is on my todo list. But I need to understand how to reach this code first. Then refactoring is going to be easier.